Per-recipe adjustments to version parsing

dunn opened the pull request:
This adds a new optional `:version-regexp` key to recipes; it's a regular expression where the capture group is the string you want sent to `version-to-list`.  I added one to the Erlang recipe as an example.

dunn commented on package-build.el:
```diff
@@ -166,55 +192,27 @@ function for access to this function")
-         (times (mapcar 'package-build--parse-time
-                        (mapcar 'cadr (package-build--string-match-all regex text 1)))))
-    (car (nreverse (sort times 'string<)))))
+(defun package-build--find-version-newest (regex &optional bound)
```
This function replaces `package-build--find-tag-version-newest` and `package-build--find-parse-time-latest`.

milkypostman replied:
Still double checking the logic here to remember how I setup this time parsing business (it's been a while) but I think you can remove `package-build--find-parse-time` too.

milkypostman replied:
ok, I see my confusion here. alright. so here is my curiosity; can we change this function so that it returns a tuple `('tag name' (0 1 2))` which is essentially `(<original string> <version parsed out of string>)`.

Because *then* we don't need to worry about re-parsing in the `package-build-package` function. it should also simplify things.

and it's not hard to change. Anything that uses this function (e.g., cvs and svn) checkouts, can just grab the `cadr`? What do you think?

i think this is close enough that if you don't want to change it up I am willing to patch in and fix up, but I'm more than happy to have your help  :)

dunn commented:
Tested by building all the A recipes in stable mode, and most of them in unstable; I didn't see any changes in versioning.

milkypostman commented:
this is amazing! Thanks! Will take a look at this in the next couple days.

milkypostman commented on README.md:
```diff
@@ -183,6 +183,9 @@ the following form (`[...]` denotes optional or conditional values),
  :fetcher [git|github|gitlab|bitbucket|bzr|hg|darcs|fossil|svn|cvs|wiki]
  [:url "<repo url>"]
  [:repo "github-gitlab-or-bitbucket-user/repo-name"]
+ [:commit "<ref>"]
```
can you remove `:commit` and `:branch` here because they are kind of hacks that I actually wish we would remove. I don't want to advertise them. there are cases where we have made exceptions to this rule but in general we avoid these things like the plague. If there is ever a time that no recipe uses either I will remove the functionality.

milkypostman commented on package-build.el:
```diff
@@ -501,17 +497,18 @@ Return a cons cell whose `car' is the root and whose `cdr' is the repository."
-            (package-build--update-git-to-ref dir (concat "tags/" (car tag-version)))
-            (cadr tag-version))
+        (let* ( (bound (goto-char (point-max)))
+                (regexp (or (plist-get (package-build--config name) :version)
```
don't need `package-build--config` here because the `config` variable should already contain the config hash. 

thus you can use `(plist-get config :version)`.

milkypostman commented on package-build.el:
```diff
@@ -769,7 +781,10 @@ Optionally PRETTY-PRINT the data."
                 (unless (symbolp (car elt))
                   (error "Invalid package name in dependency: %S" (car elt)))
-                (list (car elt) (version-to-list (cadr elt))))
+                (list
```
I don't actually think this line should use the new regexp. The regexp is for *our* packages but the dependencies shouldn't be effected by the regexp.

dunn replied:
Oh, right.

milkypostman commented on README.md:
```diff
@@ -183,6 +183,9 @@ the following form (`[...]` denotes optional or conditional values),
  [:repo "github-gitlab-or-bitbucket-user/repo-name"]
+ [:commit "<ref>"]
+ [:branch "<branch-name>"]
+ [:version "<regexp>"]
```
i think maybe we should pick something like `:version-regexp` or `:versionregexp` as a name, because `:version` makes me fear we're going to get some PRs where people will put in `2.5.4` or something.

dunn replied:
Oh, yeah, OK. `:version-regexp` sounds good.

milkypostman commented on package-build.el:
```diff
@@ -1080,9 +1094,10 @@ and a cl struct in Emacs HEAD.  This wrapper normalises the results."
     (package-build--message "\n;;; %s\n" file-name)
     (let* ((version (package-version-join
-                     (version-to-list
+                     (package-build--version-to-list
```
I don't think we need `package-build--version-to-list` here either. Because the *checkout* should return us a valid version and we should error if it doesn't. Because that means the regexp is broken.

milkypostman replied:
OK, leaving the previous comment but I get why you are doing this here. I think we should drop it based on my comments above. The *checkout* functions should basically get the tag name and check that tag out, but should still return a valid version. I would prefer that the rest of code not ever need to worry about the version regexp.

dunn replied:
Right now I've got the checkouts (at least git and hg) returning the original version-string.  `package-build--find-version-newest` uses the version list internally to figure out which tag is newest, but then returns the original tag as a string so git and hg can actually check out that release.  But if SVN, darcs, etc., return the actual list, then I need to change this.

Should `package-build--find-version-newest` return something like `'((0 1) . "v0.1")`?  Then git/hg have access to the tag string for checkouts, but the checkout function itself could just return the `car`.

dunn replied:
Oops, didn't refresh before posting.

milkypostman replied:
Yes exactly. It took me a second to figure out why we had the two functions to begin with and I'm still not sure I know but they're getting consolidated anyhow. I'm not sure how it'll work though with something like hg where you kinda need two regex. One to pull the tag and one to pull the version from the tag. However I think we can hardcore the "tag" regex and then use version-regexp to parse the version from the tag. ?

milkypostman commented on package-build.el:
```diff
@@ -166,55 +192,27 @@ function for access to this function")
+                  (or bound (point-min)) (point)))))
+    (setq tags (cl-mapcar
+                 (lambda (tag)
+                   (if (package-build--valid-version-string tag regex)
```
if -> when

milkypostman commented on package-build.el:
```diff
@@ -153,6 +159,26 @@ function for access to this function")
+(defun package-build--valid-version-string (str regexp)
+  "Parse STR with REGEXP.
+Return the result of `package-build--version-to-list', or nil if invalid"
+  (ignore-errors (package-build--version-to-list str regexp)))
```
Since I don't hink we need `package-build--version-to-list` as an independent function, you can inline it here.

milkypostman commented:
THANKS FOR THIS PR! I added a number of comments but I really wanted to say this is a nice patch and great work. Just needs a bit of polish I think. Sorry if any comments come off short or anything I appreciate your work.

dunn commented:
No problem! Thank you for the review!

dunn commented:
OK, added the changes as a new commit to make reviewing easier, I can combine them when it's ready to merge.

Other commits:
- I made `tests/version-tests.el` to check all packages for version changes. This might not be useful in the future, so feel free to drop it.
- a few packages have tags like` v.20150303`, which now gets parsed as `(0 v.20150303)` without special handling.  We could instead change the default regexp to `"^[rRvV]?\\.?\\(.*\\)$"`, but that would cause `.5` to be parsed as `(5)` instead of `(0 5)`.  Is there a way to check for `[rRvV] or [rRvV]\\.` without using a capture group?
- fixed yatex's version parsing; previously the only valid tag was `v0`.

Five other packages will have different versions after this is merged:
- **cfengine-code-style** is now `(3 8 0)` instead of `(3 8 0 2 1)`, which is correct: https://github.com/cfengine/core/releases. I think it was previously picking up one of the beta releases.
- **ecb** is now `(2 -4 24)` instead of `(2 -3 24)`.  "Snapshot" has always been parsed as `-4` so I'm not sure how it was getting `-3`.
- **epl** is now `(0 7)` instead of `(71 4 8098)`.  I think this is right, and we should drop a few :+1:s into https://github.com/cask/epl/issues/30#issue-121963688.
- **roguel-ike** is now `(0 1)` instead of `(0 1 0)`.  Those are equivalent according to `version=`; but there's a tag for each so I think it's basically random which one gets picked.
- **tablist** doesn't have any stable versions: https://github.com/politza/tablist/issues/2; was there previously a `0.60` tag? How did the current stable package get created?

dunn commented on package-build.el:
```diff
@@ -163,58 +184,31 @@ function for access to this function")
+                   (when (package-build--valid-version-string tag regex)
+                     ;; Returns a list like ((0 1) ("v0.1")); the
+                     ;; first element is used for determining the
+                     ;; newest version and the second (the original
```
Forgot to update the comment here; will do.

dunn replied:
I _did_ update this comment, but I must have lost it in a rebase :see_no_evil:

steckerhalter commented:
@milkypostman I saw you have made a comment about removing :commit and :branch from package-build. Since quelpa uses these and wants users to be able to use them in the future as well, do you think there is a way to keep them in some way?

milkypostman commented:
Yeah I don't think it's bad to keep them but I definitely don't want to
document them or endorse them for our recipes.

We have to comment a lot on cls using branch.
On Mon, Dec 14, 2015 at 12:37 steckerhalter <notifications@github.com>
wrote:

> @milkypostman <https://github.com/milkypostman> I saw you have made a
> comment about removing :commit and :branch from package-build. Since quelpa
> uses these and wants users to be able to use them in the future as well, do
> you think there is a way to keep them in some way?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/3362#issuecomment-164551828>.
>

vyp commented:
@milkypostman Yep that's fine, I assume you are okay with mentioning these in the quelpa readme?

steckerhalter commented:
@milkypostman sounds good. thanks a bunch

milkypostman commented:
Yeah np mentioning in quelpa.
On Mon, Dec 14, 2015 at 14:08 steckerhalter <notifications@github.com>
wrote:

> @milkypostman <https://github.com/milkypostman> sounds good. thanks a
> bunch
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/3362#issuecomment-164575347>.
>

purcell commented:
Yeah, we should keep mentioning them in `package.el` docstrings if applicable, but not in the MELPA README.

milkypostman commented on package-build.el:
```diff
@@ -589,17 +581,33 @@ Return a cons cell whose `car' is the root and whose `cdr' is the repository."
-            (package-build--run-process dir "hg" "update" (nth 2 tag-version))
-            (cadr tag-version))
+        (let ( (bound (goto-char (point-max)))
+               (regexp (or (plist-get (package-build--config name) :version-regexp)
```
still think you can remove the `package-build--config` here. please use the `config` argument directly.

dunn replied:
Woops, thought I fixed that.  Maybe got lost in a rebase.

dunn replied:
Oh, I changed the one in `package-build--checkout-git` and forgot to do the same here.

steckerhalter commented:
@dunn I tried the changes made in this PR. so `package-build-checkout` is now returning a list instead of a string like before, is that correct?

dunn commented:
@steckerhalter Yeah; does that break anything?

steckerhalter commented:
@dunn kind off... in quelpa we use the version string to do some more things after the checkout. I tested with your changes and it failed so that's how I noticed. I think it shouldn't be a problem to change it on our side, we would just be glad to know when it is merged so that we can push out the change.

dunn commented:
Either way works for me; the version-as-list or the version string.

dunn commented:
Looks like I didn't test non-stable well enough:
```
 • Building recipe applescript-mode ...
emacs --no-site-file --batch -l package-build.el --eval "(let ((package-build-stable nil) (package-build-write-melpa-badge-images t) (package-build-archive-dir (expand-file-name \"./packages\" package-build--this-dir))) (package-build-archive 'applescript-mode))"

;;; applescript-mode

Fetcher: svn
Source: http://svn.osdn.jp/svnroot/macemacsjp/applescript-mode/trunk

Cloning http://svn.osdn.jp/svnroot/macemacsjp/applescript-mode/trunk to /Users/cat/Documents/clones/melpa/working/applescript-mode/
File: /Users/cat/Documents/clones/melpa/packages/applescript-mode-2015snapshot12snapshot15.entry
Built in 1.707s, finished at Tue Dec 15 11:27:54 2015
 ✓ Wrote 40 -rw-r--r--  1 cat  staff    16K Dec 15 11:27 ./packages/applescript-mode-2015snapshot12snapshot15.el
 8 -rw-r--r--  1 cat  staff   247B Dec 15 11:27 ./packages/applescript-mode-2015snapshot12snapshot15.entry
 8 -rw-r--r--  1 cat  staff   775B Dec 15 11:27 ./packages/applescript-mode-badge.svg
 8 -rw-r--r--  1 cat  staff    17B Dec 15 11:27 ./packages/applescript-mode-readme.txt
```

dunn commented:
New commit that I think fixes SVN and CVS, but all of the CVS repos were timing out for me so if someone else wants to check that'd be :100:

milkypostman commented on package-build.el:
```diff
@@ -589,17 +582,33 @@ Return a cons cell whose `car' is the root and whose `cdr' is the repository."
-                                        name)))))
-            (package-build--run-process dir "hg" "update" (nth 2 tag-version))
-            (cadr tag-version))
+        (let ( (bound (goto-char (point-max)))
```
What am I missing here: we set the `bound` variable by going to the max point which I assume will return `(point-max)`. But then below we call `goto-char` again?

dunn replied:
I think what's happening is 
- we set `bound` and point to `(point-max)`
- `((package-build--run-process dir "hg" "tags")` inserts the output of `hg tags` into the buffer before point.  At this stage point is again at `(point-max)`.
- then we need to move point back to `bound` in order to do the find-and-replace

milkypostman replied:
Oh of course yes.
On Wed, Dec 16, 2015 at 09:33 Alex Dunn <notifications@github.com> wrote:

> In package-build.el
> <https://github.com/milkypostman/melpa/pull/3362#discussion_r47806527>:
>
> > @@ -589,17 +582,33 @@ Return a cons cell whose `car' is the root and whose `cdr' is the repository."
> >          (package-build--princ-checkout repo dir)
> >          (package-build--run-process nil "hg" "clone" repo dir)))
> >        (if package-build-stable
> > -          (let* ((bound (goto-char (point-max)))
> > -                 (tag-version (and (package-build--run-process dir "hg" "tags")
> > -                                   (or (package-build--find-tag-version-newest
> > -                                        "^\\(?:v[.-]?\\)?\\([0-9]+[^ \t\n]*\\)[ \t]*[0-9]+:\\([[:xdigit:]]+\\)$"
> > -                                        bound
> > -                                        2)
> > -                                       (error
> > -                                        "No valid stable versions found for %s"
> > -                                        name)))))
> > -            (package-build--run-process dir "hg" "update" (nth 2 tag-version))
> > -            (cadr tag-version))
> > +        (let ( (bound (goto-char (point-max)))
>
> I think what's happening is
>
>    - we set bound and point to (point-max)
>    - ((package-build--run-process dir "hg" "tags") inserts the output of hg
>    tags into the buffer before point. At this stage point is again at
>    (point-max).
>    - then we need to move point back to bound in order to do the
>    find-and-replace
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/3362/files#r47806527>.
>

dunn replied:
It was my fault, I didn't fully understand how the buffer manipulation worked.

steckerhalter commented:
@dunn then it would be simpler for us to keep the behavior as it was (with strings)

dunn commented:
New commit:
- `package-build-checkout` returns the version as string instead of the raw list
- I fixed some changes I made to docstrings, since I was confused about how buffer manipulation worked during the checkouts.  I saw that `bound` was before point but docstrings referred to searching "up to `bound`", and I thought it was a mistake.  I didn't realize it meant searching backwards.

milkypostman commented:
But searching backwards is ok I think right?

I will have a look tomorrow. Taking the night off from the keyboard.
On Wed, Dec 16, 2015 at 19:56 Alex Dunn <notifications@github.com> wrote:

> New commit:
>
>    - package-build-checkout returns the version as string instead of the
>    raw list
>    - I fixed some changes I made to docstrings, since I was confused
>    about how buffer manipulation worked during the checkouts. I saw that
>    bound was before point but docstrings referred to searching "up to
>    bound", and I thought it was a mistake. I didn't realize it meant
>    searching backwards.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/3362#issuecomment-165331114>.
>

dunn commented:
> But searching backwards is ok I think right?

Yep! So I restored that behavior.

dunn commented:
`blank-mode` is getting build with an older timestamp than is on MELPA:
```
Test persistent-versions condition:
    (ert-test-failed
     ((should
       (equal
        (version-to-list ...)
        (package-desc-version ...)))
      :form
      (equal
       (20130824 459)
       (20141029 816))
```
But the last-modified time in the headers is from 2013, not 2014:
```
HTTP/1.1 200 OK
Server: nginx/1.4.6 (Ubuntu)
Date: Thu, 17 Dec 2015 22:24:41 GMT
Content-Type: text/plain; charset=utf-8
Connection: keep-alive
Vary: Accept-Language
Last-Modified: Sat, 24 Aug 2013 11:59:25 GMT
X-EmacsWiki-Source: Apache/CGI
X-Proxy-Cache: HIT
```

dunn commented on package-build.el:
```diff
@@ -1096,7 +1119,7 @@ and a cl struct in Emacs HEAD.  This wrapper normalises the results."
                   (time-to-seconds (time-since start-time))
                   (current-time-string))
-      file-name)))
+      (list file-name version))))
```
This change is only for the tests, so if those are rejected this should be too (which is why they're in the same commit).

dunn commented on package-build.el:
```diff
@@ -463,11 +455,30 @@ Return a cons cell whose `car' is the root and whose `cdr' is the repository."
-            (error "No valid timestamps found!"))
-        ))))
 
+        ;; `cvs log` does not provide a way to view the previous N
```
I realized the `package-build--find-parse-time-latest` function was still being used by the CVS checkout function, so that's why I added this.

milkypostman commented on package-build.el:
```diff
@@ -153,6 +159,21 @@ function for access to this function")
 
+
+
+(defun package-build--config (package)
```
sorry i keep harping on this but this function should go away and the code should just live in the one place where it is used.

dunn replied:
Oh no problem, I didn't realize it was down to one instance.

milkypostman commented:
added one last comment and I have one more thing to ask (even though I do feel slightly bad asking it given how much you have done): could you squash these commits into two or three such that they are,

- made the changes to package-build to support version regex
- updated recipes


Then when you are done with your final push, I will take and run this against all the recipes just to see that we are the same on MELPA as we are now.

Then I will take and do the same with MELPA stable.  I will spin up a clone of our servers and just run it there (with your patch of course) to confirm it's not broken anything.

milkypostman commented:
Also, if you are busy and all that I can do the squashing and reworking of the commits, it's just easier if I can hit merge here after I verify everything.

dunn commented:
The recipe changes shouldn't be one commit per recipe?

milkypostman commented:
they don't need to be. that's fine if you want to. mainly I just wanted to clean up all the commits to package-build. but you can collapse the recipe changes too if youw ant.

dunn commented:
If you're fine either way I'll leave the recipe commits separate.  What about the ERT tests; do you want those or should I drop them?

milkypostman commented:
i don't think so!  More tests the better!!!

dunn commented:
OK, squashed!

milkypostman commented:
Will get to this tomorrow testing. Will need some time don't have it
tonight.
On Thu, Dec 17, 2015 at 21:37 Alex Dunn <notifications@github.com> wrote:

> OK, squashed!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/3362#issuecomment-165678315>.
>

purcell commented:
Thanks @dunn, and thanks @milkypostman for helping with all this -- I'm pretty slammed at the moment.

milkypostman commented:
@dunn looks good on my test server! i will submit first thing tomorrow morning. don't want to sleep through the night and have anything accidentally break.

milkypostman commented:
THANKS! WE MADE IT!!!

purcell commented:
Awesome work folks. Thanks so much for all your @dunn!

dunn commented:
No problem! Let me know if anything breaks down the line.

mdorman commented:
I notice, though, that epl on melpa-stable still has the wonky version number that's mentioned near the beginning of this discussion.  Do those (and perhaps the other packages that got recipe tweaks) need to be manually kicked to get them corrected?

dunn commented:
@mdorman yeah, looks like it. cc @milkypostman, the packages whose versions dropped are https://github.com/milkypostman/melpa/pull/3362#issuecomment-164361596

milkypostman commented:
@mdorman i don't think they are "wonky" just that they will have different numbers now, but @dunn they are correct right? or do we need to update the regex?

dunn commented:
The problem is that a few packages now have lower versions than previously, like EPL: http://stable.melpa.org/#/epl

And they aren't being made available since the build scripts think they're older.

milkypostman commented:
this I can fix. later tonight I will cause a complete rebuild. some
packages will be missing for a while but *shrug* its currently the only way
to make things right in the world. sorry I don't have a better answer but I
haven't spent much time in the stable world.


On Tue, Jan 19, 2016 at 4:56 PM Alex Dunn <notifications@github.com> wrote:

> The problem is that a few packages now have lower versions than
> previously, like EPL: http://stable.melpa.org/#/epl
>
> And they aren't being made available since the build scripts think they're
> older.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/3362#issuecomment-173038435>.
>

dunn commented:
That's fine, thanks so much!

milkypostman commented:
everything rebuilt. do the versions look as you expect?

dunn commented:
All perfect except http://stable.melpa.org/#/cfengine-code-style; when I build stable locally it's 3.8.0 which I think is what it's supposed to be: https://github.com/cfengine/core/releases
